src/sync/patch.rs

```diff
@@ -1,66 +1,50 @@
 use crate::db;
 use crate::util::rlog;
 use serde::Deserialize;
-use serde_json::json;
-use std::default::Default;
 
-const OP_ADD: &str = "add";
-const OP_REMOVE: &str = "remove";
-const OP_REPLACE: &str = "replace";
-
-#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
-pub struct Operation {
-    pub op: String,
-    pub path: String,
-    #[serde(default)]
-    pub value: serde_json::Value,
+#[derive(Deserialize)]
+#[cfg_attr(test, derive(Clone, Debug, PartialEq))]
+#[serde(tag = "op")]
+pub enum Operation {
+    #[serde(rename = "put")]
+    Put {
+        key: String,
+        value: serde_json::Value,
+    },
+    #[serde(rename = "del")]
+    Del { key: String },
+    #[serde(rename = "clear")]
+    Clear,
 }
 
 pub async fn apply(db_write: &mut db::Write<'_>, patch: &[Operation]) -> Result<(), PatchError> {
     use PatchError::*;
     for op in patch.iter() {
-        // Special case `{"op": "replace", "path": "", "value": {}}`
-        // which means replace the root with a new map, in other words, clear
-        // the map.
-        if op.path.is_empty() {
-            if op.op == OP_REPLACE && op.value == json!({}) {
-                db_write.clear().await.map_err(ClearError)?;
-                continue;
-            }
-            return Err(InvalidPath(op.path.clone()));
-        }
-
-        let mut chars = op.path.chars();
-        if chars.next() != Some('/') {
-            return Err(InvalidPath(op.path.clone()));
-        }
-        let key = json_pointer_unescape(chars.as_str()).as_bytes().to_vec();
-
-        match op.op.as_str() {
-            OP_ADD | OP_REPLACE => {
-                let value = serde_json::to_vec(&op.value).map_err(InvalidValue)?;
+        match op {
+            Operation::Put { key, value } => {
+                let key = key.as_bytes().to_vec();
+                let value = serde_json::to_vec(value).map_err(InvalidValue)?;
```

Comment:
@aboodman note that our policy is not error out if all there is a problem with any op. I think this is right because we don't want to partially apply an update. Just calling it out.

Reply:
To be clear the policy is to exit early in case of error and abort the whole patch, right? If so, lgtm.

Reply:
I added a test to enforce that in the follow up.